design-system/src/theme-provider.tsx

```diff
@@ -54,41 +29,74 @@ const validateTheme = (themeName?: string): ThemeName => {
   return 'default';
 };
 
-const ThemeProvider = (props: ThemeProviderProps) => {
-  const root = useRef<HTMLElement>(document.querySelector(':root'));
-  const localScopeElement = useRef<HTMLDivElement>(null);
-  const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
+// used to cover SSR builds (for instance in Gatsby)
+const isBrowser = typeof window !== 'undefined';
 
-  const changeTheme = useCallback((newTheme: string) => {
-    setTheme(validateTheme(newTheme));
-  }, []);
+const defaultParentSelector = (): HTMLElement | null =>
+  document.querySelector(':root');
 
-  useLayoutEffect(() => {
-    const vars = toVars(themes[theme]);
-    const targetElement =
-      props.scope === 'local' ? localScopeElement.current : root.current;
-    Object.entries(vars).forEach(([key, value]) => {
-      targetElement?.style.setProperty(key, value);
-    });
-  }, [theme, props.scope]);
+type TChangeTheme = {
+  newTheme?: string;
+  parentSelector: typeof defaultParentSelector;
+  themeOverrides?: Record<string, string>;
+};
 
-  const value = useMemo(() => {
-    return { theme, changeTheme };
-  }, [theme, changeTheme]);
+const changeTheme = ({
+  newTheme,
+  parentSelector = defaultParentSelector,
+  themeOverrides,
+}: TChangeTheme): void => {
+  const target = isBrowser ? parentSelector() : null;
+  const validTheme = validateTheme(newTheme);
+  target?.setAttribute('data-theme', validTheme);
```

Comment:
The idea behind this solution is:
- adding a `data-theme` either to `:root` or the parent based on the provided selector function. `data-theme` stores the current theme
-  adding all the custom properties to `style`

On each change of `theme`, `customProperties` or `parentSelector` the theme change is triggered within `useLayoutEffect`.